Add Menu-based Actions

ataulm opened the pull request:
This is to address #27.

This PR adds a **menu-based solution** for exposing item view actions to accessibility services, specifying actions in `res/menu/actions.xml` and providing a `MenuItem.OnMenuItemClickListener`.

---

An `Action` is an abstraction around a callback available for a given item view in a list, where each action is composed of:

- an ID
- a label
- a runnable to execute when the action is triggered

![talkback local context menu showing custom actions](https://blog.novoda.com/content/images/2016/11/07.png)

The abstraction was created to group these three things together in such a way that they could be used in an accessibility delegate to expose these callbacks to accessibility services like TalkBack.

See [this blog post](https://blog.novoda.com/accessibilitools/) for more information.

---

The API for creating actions is pretty hefty and can result in a lot of unfamiliar boilerplate in the classes where views are bound.

**This PR aims to reduce the overhead of learning this API by co-opting the Android Menu paradigm: specifying actions in `res/menu/actions.xml` and providing a `MenuItem.OnMenuItemClickListener`.**

The size of this PR is mostly due to the first two classes, `ActionsMenu implements Menu` and `ActionMenuItem implements MenuItem`. These are package-protected, and I've only implemented as much as is necessary to make this work.

---

If this approach is amenable, I would recommend deprecating the following classes:

- `Action`
- `Actions`
- `ActionsAlertDialogCreator`
- `ActionsAccessibilityDelegate`

ataulm committed:
Add ActionMenuItem and ActionsMenu

ataulm committed:
Add ActionsMenuAccessibilityDelegate

ataulm committed:
Add menu-based implementations of Actions*AccessibilityDelegate and ActionsMenuAlertDialog

ataulm committed:
Update demo TweetView with menu.xml based actions

ataulm committed:
Lift MenuInflator to field to avoid a new instance for each usage of inflate

ataulm committed:
Add blank line to EOF

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+import android.view.SubMenu;
+import android.view.View;
+
+class ActionMenuItem implements MenuItem {
```
In both this and `ActionsMenu`, I used `UnsupportedOperationException` for the methods that were:

- we don't use/need them
- they're not called by `MenuInflater` during inflation

ataulm replied:
*were 🙀

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    @Override
+    public ActionMenuItem setTitleCondensed(@Nullable CharSequence title) {
```
`MenuInflater` calls this with `null` 🙄 so I wasn't able to throw an exception indiscriminately - there are others like this.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    @Override
+    public ActionMenuItem setIcon(Drawable icon) {
```
We could eventually support this and the next - perhaps the AlertDialog we show could be styled or have a custom item view which displays the icon **and** title.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+
+    @Override
+    public ActionMenuItem setNumericShortcut(char numericChar) {
+        Log.w("!!!", "setNumericShortcut(char) is not supported but is called by MenuInflater");
```
`MenuInflater` calls this and the Alphabetic one with some rando character - this is a no-op, but because we're not throwing the exception, I put a warning log to highlight it in some way to consumers.

jszmltr replied:
👍 I would change the tag tho'

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    @Override
+    public ActionMenuItem setChecked(boolean checked) {
```
`MenuInflater` calls this without checking if it's checkable 🤔 it calls it with `false` I guess to set a default value.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+        return false;
+    }
+
+    @Nullable
```
I checked the `MenuItem` interface and only return null for methods where it says it's ok.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionsMenu.java:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Override
+    public void setGroupCheckable(int group, boolean checkable, boolean exclusive) {
```
we could support `setGroupCheckable` at this level, ignoring whether `ActionMenuItem` supports them but will leave for another PR.

I would have done it except for the `exclusive` flag, which makes it a bit annoying to do it properly.

ataulm committed:
Add support for setGroupVisible and setGroupEnabled

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionsMenuAccessibilityDelegate.java:
```diff
@@ -0,0 +1,39 @@
+        if (item == null) {
+            return super.performAccessibilityAction(host, actionId, args);
+        } else {
+            menuItemClickListener.onMenuItemClick(item);
```
I dunno if this is legit or if we should be casting to `ActionMenuItem` and calling `invokeAction()`, which calls `onMenuItemClick(item)` anyway.

If we do it this way, then at least, we can pass _any_ menu to this class; we don't need to expose `ActionsMenu`.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionsMenuAlertDialog.java:
```diff
@@ -0,0 +1,24 @@
+
+public class ActionsMenuAlertDialog {
+
+    public static AlertDialog.Builder create(Context context, Menu menu, MenuItem.OnMenuItemClickListener menuItemClickListener) {
```
returning the builder instead gives the consumer a chance to set the title or not, and other things if they want.

ataulm commented on demo/src/main/java/com/novoda/accessibility/demo/custom_actions/TweetView.java:
```diff
@@ -44,23 +43,43 @@ protected void onFinishInflate() {
+        MenuItem.OnMenuItemClickListener menuItemClickListener = createMenuItemClickListener(tweet, listener);
+        Menu menu = actionsMenuInflater.inflateMenu(R.menu.tweet_actions, menuItemClickListener);
+        ActionsMenuAccessibilityDelegate delegate = new ActionsMenuAccessibilityDelegate(menu, menuItemClickListener);
+//      TODO: add composable delegate - delegate.setClickLabel(R.string.tweet_actions_usage_hint);
```
not sure about this bit yet - I was going to add `UsageHintsAccessibilityDelegate` like in `ActionsAccessibilityDelegate` but it doesn't seem very scaleable.

I wonder if it's better to take a collection of `AccessibilityDelegate` and have a `CompositeAccessibilityDelegate` or some sort.

ataulm replied:
On looking at this, I'm not sure it's possible to do this - at least it's not immediately obvious how a CompositeAccessibilityDelegate would behave.

There are many methods inside AccessibilityDelegate and it's unclear whether it's appropriate to call _each_ of the delegates or just first one that accepts the event (some methods return boolean).

For this reason, I think it'd be most appropriate if we either let the user define their own composite delegate, in which they can call to the UsageHints one and the `ActionsMenuAccessibilityDelegate` OR add `UsageHints` behaviour to the `Actions` one as we have already in `ActionsAccessibilityDelegate`.

I'd recommend the former for now, since we'll have the chance to add it later, whereas adding it now will mean we might have to remove it later which is more pain.

ataulm commented on core/src/main/java/com/novoda/accessibility/ActionsMenuInflater.java:
```diff
@@ -0,0 +1,32 @@
+    private final MenuInflater menuInflater;
+    private final Resources resources;
+
+    public ActionsMenuInflater(Context context) {
```
maybe a static method `ActionsMenuInflator.from(Context)` to match `LayoutInflater`?

ataulm replied:
commit 8eed69cb7476744bade33c805d35cdc38b98df5b (HEAD -> menu-based-actions, origin/menu-based-actions)
Author: Ataul Munim <hello@ataulm.com>
Date:   Tue Apr 3 13:14:54 2018 +0100

    Replace worky ctor with static factory method to match layoutInflater

takecare commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+     */
+    boolean invokeAction() {
+        if (isEnabled() && menuItemClickListener != null) {
+            menuItemClickListener.onMenuItemClick(this);
```
why not `return menuItemClickListener.onMenuItemClick(this);`?

ataulm replied:
commit c231b6a3156aec9897c80f5c2fbe2207bddd65ee
Author: Ataul Munim <hello@ataulm.com>
Date:   Tue Apr 3 13:12:04 2018 +0100

    Return result of menuitemclicklistener

takecare commented on core/src/main/java/com/novoda/accessibility/ActionMenuItem.java:
```diff
@@ -0,0 +1,273 @@
+    @Nullable
+    @Override
+    public ActionProvider getActionProvider() {
+        return null;
```
no `UnsupportedOperationException`?

ataulm replied:
For the ones that were nullable, I returned null - in this case it's also legit to return null to, since there is no actionview/action provider.

takecare commented on core/src/main/java/com/novoda/accessibility/ActionsMenuAccessibilityDelegate.java:
```diff
@@ -0,0 +1,39 @@
+            return super.performAccessibilityAction(host, actionId, args);
+        } else {
+            menuItemClickListener.onMenuItemClick(item);
+            return true;
```
same as above regarding the return value of `onMenuItemClick`. could delegate to super if false?

ataulm replied:
yep makes sense 👍

ataulm replied:
https://github.com/novoda/accessibilitools/pull/28#discussion_r178802623

ataulm committed:
Return result of menuitemclicklistener

ataulm committed:
Replace worky ctor with static factory method to match layoutInflater

ataulm commented on demo/src/main/java/com/novoda/accessibility/demo/custom_actions/TweetView.java:
```diff
@@ -44,23 +43,43 @@ protected void onFinishInflate() {
+        MenuItem.OnMenuItemClickListener menuItemClickListener = createMenuItemClickListener(tweet, listener);
+        Menu menu = actionsMenuInflater.inflate(R.menu.tweet_actions, menuItemClickListener);
+        ActionsMenuAccessibilityDelegate delegate = new ActionsMenuAccessibilityDelegate(menu, menuItemClickListener);
+//      TODO: add composable delegate - delegate.setClickLabel(R.string.tweet_actions_usage_hint);
```
fully composable delegate not legit - will make one _in the demo_ that wraps this and UsageHints one

jszmltr replied:
Would rather just kill this TODO and add an issue.

ataulm commented:
:+1: both good points I will do it today

On Thu, 5 Apr 2018, 08:52 Jacek Szmelter, <notifications@github.com> wrote:

> *@jackSzm* commented on this pull request.
> ------------------------------
>
> In core/src/main/java/com/novoda/accessibility/ActionMenuItem.java
> <https://github.com/novoda/accessibilitools/pull/28#discussion_r179375244>
> :
>
> > +        throw new UnsupportedOperationException();
> +    }
> +
> +    @Override
> +    public Intent getIntent() {
> +        throw new UnsupportedOperationException();
> +    }
> +
> +    @Override
> +    public ActionMenuItem setShortcut(char numericChar, char alphaChar) {
> +        throw new UnsupportedOperationException();
> +    }
> +
> +    @Override
> +    public ActionMenuItem setNumericShortcut(char numericChar) {
> +        Log.w("!!!", "setNumericShortcut(char) is not supported but is called by MenuInflater");
>
> 👍 I would change the tag tho'
> ------------------------------
>
> In
> demo/src/main/java/com/novoda/accessibility/demo/custom_actions/TweetView.java
> <https://github.com/novoda/accessibilitools/pull/28#discussion_r179376193>
> :
>
> > @@ -44,23 +43,43 @@ protected void onFinishInflate() {
>      }
>
>      public void display(String tweet, Listener listener) {
> -        Actions actions = createActions(tweet, listener);
> -        ActionsAccessibilityDelegate delegate = new ActionsAccessibilityDelegate(getResources(), actions);
> -        delegate.setClickLabel(R.string.tweet_actions_usage_hint);
> +        MenuItem.OnMenuItemClickListener menuItemClickListener = createMenuItemClickListener(tweet, listener);
> +        Menu menu = actionsMenuInflater.inflate(R.menu.tweet_actions, menuItemClickListener);
> +        ActionsMenuAccessibilityDelegate delegate = new ActionsMenuAccessibilityDelegate(menu, menuItemClickListener);
> +//      TODO: add composable delegate - delegate.setClickLabel(R.string.tweet_actions_usage_hint);
>
> Would rather just kill this TODO and add an issue.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/novoda/accessibilitools/pull/28#discussion_r179375244>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACjfG2XOYpbAJufz_znq2IIUImXdzlPAks5tlc0ggaJpZM4TCntY>
> .
>

-- 
Disclaimer: The information in this e-mail and any attachments is the 
property of Novoda Ltd and is confidential and may be legally privileged. 
This e-mail is intended solely for the person or organisation to which it 
is addressed. Any disclosure, copying or other use of the information by 
any person or organisation who is not the intended recipient is strictly 
prohibited and may be unlawful. If you have received this e-mail in error, 
please inform the sender immediately and delete/destroy this e-mail and any 
copies of it. Novoda Ltd has taken reasonable precautions to minimise the 
risk of any software viruses which may damage your systems, but we advise 
that you take the necessary steps to ensure that no virus contamination is 
suffered. Novoda Ltd does not accept any liability for any loss or damage 
caused by the transmission of any virus.

Novoda Ltd, Company No: 347444, 
Registered in Scotland Registered Office: C/O Alexander Sloan, 38 Cadogan 
Street, Glasgow, G2 7HF, Scotland.  VAT Registration Number  GB 984 2525 93

ataulm committed:
Replace !!! with accessibilitools for warning log

ataulm committed:
Extract UsageHints to be reused between delegates

- not possible to create composed AccessibilityDelegates in an
isolated fashion (i.e. one that contains multiple others) without
the wrapper one knowing the implementation of the ones it's wrapping
- this is because calling initialize multiple times on the same node
can overwrite information set by the preceding delegate
- it worked for the ActionsAccessiblityDelegate only because the
UsageHintsDelegate was called _after_, and it happened not to
update those actions
- instead, extracted UsageHints and users will have to pass it in
explicitly, apart from UsageHintsAccessibilityDelegate which just
mirrors the API

ataulm commented:
@jackSzm I changed base to `develop` since my latest commit will affect the API, and not sure when it'll be released.

commit cb36c9e7710474a237efb2e25a7ade4655f72df5
Author: Ataul Munim <hello@ataulm.com>
Date:   Thu Apr 5 15:06:07 2018 +0100

    Replace !!! with accessibilitools for warning log


commit aed66d4a2dfa57215111096895fd5e687c1da5b8
Author: Ataul Munim <hello@ataulm.com>
Date:   Thu Apr 5 15:59:42 2018 +0100

    Extract UsageHints to be reused between delegates
    
    - not possible to create composed AccessibilityDelegates in an
    isolated fashion (i.e. one that contains multiple others) without
    the wrapper one knowing the implementation of the ones it's wrapping
    - this is because calling initialize multiple times on the same node
    can overwrite information set by the preceding delegate
    - it worked for the ActionsAccessiblityDelegate only because the
    UsageHintsDelegate was called _after_, and it happened not to
    update those actions
    - instead, extracted UsageHints and users will have to pass it in
    explicitly, apart from UsageHintsAccessibilityDelegate which just
    mirrors the API
